src/main/java/picard/illumina/ExtractBarcodesProgram.java

```diff
@@ -0,0 +1,155 @@
+    public File OUTPUT_DIR;
+
+    @Argument(doc = "Per-barcode and per-lane metrics written to this file.", shortName = StandardOptionDefinitions.METRICS_FILE_SHORT_NAME, optional = true)
+    public File METRICS_FILE;
+
+    @Argument
+    public Boolean BARCODE_EXTRACT = false;
+
+    @Argument(optional = true)
+    public File INPUT_PARAMS_FILE;
+
+    protected IlluminaDataProviderFactory factory;
+    protected final Map<String, BarcodeMetric> barcodeToMetrics = new LinkedHashMap<>();
+    protected final static int MAX_LOOKUP_SIZE = 4096;
+    protected final ConcurrentHashMap<BarcodeExtractor.ByteString, BarcodeExtractor.BarcodeMatch> barcodeLookupMap = new ConcurrentHashMap<>(MAX_LOOKUP_SIZE);
+    protected final Set<BarcodeExtractor.ByteString> barcodeByteStrings = new HashSet<>();
+    protected final BclQualityEvaluationStrategy bclQualityEvaluationStrategy = new BclQualityEvaluationStrategy(MINIMUM_QUALITY);
+    private final NumberFormat tileNumberFormatter = NumberFormat.getNumberInstance();
+
+    /**
+     * The read structure of the actual Illumina Run, i.e. the readStructure of the input data
+     */
+    protected ReadStructure readStructure;
+
+    protected BarcodeExtractor createBarcodeExtractor() {
+        // Create BarcodeMetric for counting reads that don't match any barcode
+        final String[] noMatchBarcode = new String[readStructure.sampleBarcodes.length()];
+        final byte[][] perfectScores = new byte[readStructure.sampleBarcodes.length()][];
+        int index = 0;
+        for (final ReadDescriptor d : readStructure.descriptors) {
+            if (d.type == ReadType.Barcode) {
+                perfectScores[index] = new byte[d.length];
+                Arrays.fill(perfectScores[index], (byte) 60);
+                noMatchBarcode[index++] = StringUtil.repeatCharNTimes('N', d.length);
+            }
+        }
+
+        BarcodeMetric noMatchMetric = new BarcodeMetric(null, null, IlluminaUtil.barcodeSeqsToString(noMatchBarcode), noMatchBarcode);
+
+        for (final BarcodeMetric metric : barcodeToMetrics.values()) {
+            this.barcodeByteStrings.add(new BarcodeExtractor.ByteString(metric.barcodeBytes));
+        }
+
+        final BarcodeExtractor barcodeExtractor = new BarcodeExtractor(barcodeToMetrics,
+                barcodeLookupMap,
+                barcodeByteStrings,
+                noMatchMetric,
+                MAX_NO_CALLS,
+                MAX_MISMATCHES,
+                MIN_MISMATCH_DELTA,
+                MINIMUM_BASE_QUALITY,
+                DISTANCE_MODE);
+
+        // Prepopulate the lookup map with all perfect barcodes
+        for(BarcodeMetric metric : barcodeToMetrics.values()){
+            BarcodeExtractor.BarcodeMatch match = barcodeExtractor.calculateBarcodeMatch(metric.barcodeBytes,
+                    perfectScores);
+            barcodeLookupMap.put(new BarcodeExtractor.ByteString(metric.barcodeBytes), match);
+        }
```

Comment:
Could this also be done inside the extractor?  Do we need to expose the cache at all even?